client/deviceconnect/client.go

```diff
@@ -0,0 +1,161 @@
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+package deviceconnect
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/gorilla/websocket"
+	"github.com/mendersoftware/go-lib-micro/ws"
+	"github.com/pkg/errors"
+	"github.com/vmihailenco/msgpack"
+)
+
+const (
+	// protocols
+	httpProtocol = "http"
+	wsProtocol   = "ws"
+
+	// Time allowed to write a message to the peer.
+	writeWait = 10 * time.Second
+
+	// Time allowed to read the next pong message from the peer.
+	pongWait = 1 * time.Minute
+
+	// deviceconnect API path
+	deviceConnectPath = "/api/management/v1/deviceconnect/devices/:deviceID/connect"
+)
+
+type Client struct {
+	url        string
+	skipVerify bool
+	conn       *websocket.Conn
+}
+
+func NewClient(url string, skipVerify bool) *Client {
+	return &Client{
+		url:        url,
+		skipVerify: skipVerify,
+	}
+}
+
+// Connect to the websocket
+func (c *Client) Connect(deviceID string, token []byte) error {
+	fmt.Printf("Connecting to the device %s...\n", deviceID)
+	u, err := url.Parse(strings.TrimSuffix(c.url, "/") + strings.Replace(deviceConnectPath, ":deviceID", deviceID, 1))
+	if err != nil {
+		return errors.Wrap(err, "Unable to parse the server URL")
+	}
+	u.Scheme = strings.Replace(u.Scheme, httpProtocol, wsProtocol, 1)
+
+	headers := http.Header{}
+	headers.Set("Authorization", "Bearer "+string(token))
+	websocket.DefaultDialer.TLSClientConfig = &tls.Config{
+		InsecureSkipVerify: c.skipVerify,
+	}
+	conn, _, err := websocket.DefaultDialer.Dial(u.String(), headers)
+	if err != nil {
+		return errors.Wrap(err, "Unable to connect to the device")
+	}
+
+	err = conn.SetReadDeadline(time.Now().Add(pongWait))
+	if err != nil {
+		return errors.Wrap(err, "Unable to set the read deadline")
+	}
+
+	c.conn = conn
+	return nil
+}
+
+// PingPong handles the ping-pong connection health check
+func (c *Client) PingPong(ctx context.Context) {
+	pingPeriod := (pongWait * 9) / 10
+	ticker := time.NewTicker(pingPeriod)
+	defer ticker.Stop()
+
+	c.conn.SetPongHandler(func(string) error {
+		ticker.Reset(pingPeriod)
+		return c.conn.SetReadDeadline(time.Now().Add(pongWait))
+	})
+
+	c.conn.SetPingHandler(func(msg string) error {
+		ticker.Reset(pingPeriod)
+		err := c.conn.SetReadDeadline(time.Now().Add(pongWait))
+		if err != nil {
+			return err
+		}
+		return c.conn.WriteControl(
+			websocket.PongMessage,
+			[]byte(msg),
+			time.Now().Add(writeWait),
+		)
+	})
+
+	for {
+		select {
+		case <-ticker.C:
+			pongWaitString := strconv.Itoa(int(pongWait.Seconds()))
+			_ = c.conn.WriteControl(
+				websocket.PingMessage,
+				[]byte(pongWaitString),
+				time.Now().Add(writeWait),
+			)
+
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+// ReadMessage reads a Proto message from the websocket
+func (c *Client) ReadMessage() (*ws.ProtoMsg, error) {
+	_, data, err := c.conn.ReadMessage()
```

Comment:
`c.conn.Read*` should be protected by a mutex: https://pkg.go.dev/github.com/gorilla/websocket#hdr-Concurrency

Reply:
@alfrunes  I can add the mutex, but I believe there is a single go-routine reading and writing messages using the gorilla websocket connection. Am I wrong? By the way, I simply refactored the previous implementation from the Remote Terminal into a dedicated client to reuse it in the Port Forwarding, I'm not changing the RT implementation.

Reply:
websocket ping packet is written in a separate go-routine (actually the routine calling ReadMessage). I'm not sure if the packet is so small that can be considered to be a single "atomic" write, but I'd rather be safe than sorry.

Reply:
You are right, let's do it: https://github.com/mendersoftware/mender-cli/pull/106/commits/243c0b7c73650314f9a2e6e371874fc8130bcb6a